### PR TITLE
[test-case] should trigger only once when path is changed

### PR DESCRIPTION
**Pull request by Lucifier129**

### CODE REVIEW

File: modules/__tests__/describePush.js [hunk 1]
```diff
@@ -32,6 +32,17 @@ const describePush = (createHistory) => {
...
         execSteps(steps, history, done)
       })
+
+      it('should trigger only once when path is changed', (done) => {
+        const spy = expect.createSpy()
+        const unlisten = history.listen(spy)
+        history.push('/test')
+        setTimeout(() => {
+          expect(spy.calls.length).toBe(1)
+          unlisten()
```
Comment (taion) [inline modules/__tests__/describePush.js]: Please catch errors and ensure that you call `unlisten` even if the assertion fails.

**Comment by taion** [comment]
I see now. Thanks. Do you want to move the fix onto this PR, or re-open the old one?

**Comment by Lucifier129** [comment]
already done

File: modules/__tests__/describePush.js [hunk 2]
```diff
@@ -32,6 +32,22 @@ const describePush = (createHistory) => {
...
+        const unlisten = history.listen(spy)
+        history.push('/test')
+        setTimeout(() => {
+          try {
+            expect(spy.calls.length).toBe(1)
+            done()
+          } catch(error) {
+            throw error
+          } finally {
+            unlisten()
```
Comment (taion) [inline modules/__tests__/describePush.js]: Better to just put this in an `afterEach` IMO, given that you control the test case here.

File: modules/__tests__/describePush.js [hunk 3]
```diff
@@ -32,6 +32,22 @@ const describePush = (createHistory) => {
...
+          }
+        }, 10)
```
Comment (taion) [inline modules/__tests__/describePush.js]: Do you need the explicit 10 ms timeout here? I feel like it's not ideal to hard-code timeouts when avoidable.

File: modules/HashProtocol.js [hunk 4]
```diff
@@ -61,9 +61,13 @@ export const startListener = (listener, pathCoder, queryKey) => {
       replaceHashPath(encodedPath)
     } else {
       const currentLocation = getCurrentLocation(pathCoder, queryKey)
+      const isEqualKey = prevLocation && currentLocation.key && prevLocation.key === currentLocation.key
+      const isEqualPathname = prevLocation && prevLocation.pathname === currentLocation.pathname
+      const isEqualSearch = prevLocation && prevLocation.search === currentLocation.search
 
-      if (prevLocation && currentLocation.key && prevLocation.key === currentLocation.key)
+      if (isEqualKey || isEqualPathname && isEqualSearch) {
```
Comment (taion) [inline modules/HashProtocol.js]: This logic isn't quite right. See perhaps https://github.com/ReactTraining/history/blob/f247b4001cd1ad8e477a1e131cada1cba58fc1fe/modules/createHistory.js#L100-L106.

**Comment by Aben** [comment]
``` js
let history =  createHashHistory()
history.listen(location => {
  console.log(location.pathname) //  execute twice, why???
})
// Let's follow the code
push('/foo') 
// and then invoke transitonTo
const push = (input) =>
  transitionTo(createLocation(input, PUSH))
// and then 
if (pushLocation(nextLocation) !== false)
  updateLocation(nextLocation)
// in updateLocation, listener will be executed(1) 
// Let's focus on pushLocation
if (getHashPath() !== path) {
  pushHashPath(path)
}
// hashchange event is fired, next step invoke transitionTo again(2).
const pushHashPath = (path) =>
  window.location.hash = path
```

**Comment by taion** [comment]
Why did you close this? This looks like a good fix to have.

**Comment by Aben** [comment]
I readed again.  trigger twice only `push/replace` without `state`, so  we could add `queryKey` for all path, but it's ugly, this is a good fix. @Lucifier129 @taion

**Comment by taion** [comment]
Hmm, #316 might fix this as well then.

**Comment by taion** [comment]
I'm going to try to pull this test case into #316.

**Comment by taion** [comment]
Done – and this is in fact fixed in #316

**Pull request closed without merging**